### DISCUSSION TITLE
Syllable Stress

### DISCUSSION THREAD

**Discussion by louderpages** [RHVoice/RHVoice#441] [Show and tell]
Hi @grzezlo and @zstanecic,

As I understand it, you need to implement syllable stress in your Polish fomas.  This does not involve the pitch code, it is done by marking vowels with a "1".

Forgive me if I get the line numbers wrong - I don't have time to refer to versions.

Macedonian g2p.foma line 107 defines antePenultimateStressRule.  Look at lines 55 and 56 where unstressed and stressed versions of vowels are defined.  You can use that rule for Polish and add another for monosyllablc words. You may need a small dictionary for  irregular and foreign loan-words, as has been done with Macedonian.

Good luck!

**Comment by louderpages** [comment]
So sorry. Lines 71 and 72, not 55, 56.  At least in my local version. We are working on an upgrade to Macedonian

**Comment by zstanecic** [comment]
And one more thing:

The pitch code, which we have is borrowed from Macedonian, it worked in the non-data only version, and we want to fix this, too.

We don’t use the eos flag in our implementation, I hope that it’s not needed.

Best,

Zvonimir

**Comment by Olga-Yakovleva** [comment]
@zstanecic, the pitch situation is not related to the topic of this discussion. Open an issue about it and give me detailed instructions of how to get your voices in order to reproduce it, as they aren't yet in our repository.

**Comment by zstanecic** [comment]
Hi Olga,

#442

**Comment by Olga-Yakovleva** [comment]
Actually, I think my comment was wrong. Only the problem is in your rules which don't assign stress yet. If you copied the pitch modificationtree from Macedonian as is, those rules will not work if you don't assign stress, as they use that information. Use the information @louderpages has given you here to fix that first. Check that the stress is being assigned correctly using the RHVoice-transcribe-sentences command with the -v or -s flag. Than let me know if you still have problems with prosody.

**Reply by grzezlo** [reply]
@Olga-Yakovleva, I'm sure that putting stress will solve the problem, because when the stress has been set in code in  post_lex function, intonation for questions and exclamations worked fine.

**Comment by grzezlo** [comment]
Thanks @louderpages for your precious advice.
I was sure, that post_lex function has been replaced by the pg2p, had no 
idea, that on the g2p stage we can set stress on vowel putting 1 after.
I'll try to understand the foma magic in places pointed by you and adapt 
it to our language.

**Comment by zstanecic** [comment]
Thanks once more for the confirmation.

I’ve opened the issue, as you asked. Thanks Olga for and sorry for bothering.

**Comment by louderpages** [comment]
@grzezlo, On pg2p and stress rules.

You may need pg2p if you have consonant assimilation across word boundaries in Polish, like in a lot of Slavic languages.  You know, if you have say, a B at the end of word to join to a “T” in the next word, so the T maybe be pronounced like a “D”.

Anyway, in Macedonian this is necessary and requires that the foma looks at two words, not just the one, like with g2p.  Actually, thinking back, I think that HTS learned to do this assimilation from the voice recordings anyway, but there are special words in Macedonian which. if they occur at the end of a sentence. do not follow the assimilation.  Well, not so simple as that, but that is the general issue, and is why there is an “eos” - end of sentence - marker in the tokenizer.

CORRECTION:  We did not use the e.o.s. in pg2p. I think it was in order to handle abbreviations which include a period.  Not sure.   So many things going on. 

I should have pointed out earlier that the Macedonian standard stress rule is antepenultimate (oh, the Latin!). In English, the one before the one before the end. Or “third from the end”. Whereas in Polish, I think you want just penultimate - second from the end. But you will see in that Macedonian rule how to change it to do that.

Consider yourself lucky that you have such regular stress in Polish.  English is crazy, but perhaps not even as wierd as Croatian.  What are those crazy Croatians up to? (Smile)

- louderpages

**Comment by zstanecic** [comment]
Crazy croatians and serbs. Haha. 

Accent system with four types of rising and falling accents. 

That’s why I don’t want to write rhvoice module for Croatian. Too big dictionary will be needed.

**Comment by grzezlo** [comment]
Unfortunately, only partial success at the moment.
My stress rules work in foma software, but not in transcribe-sentences. No idea what did I wrong.

example:

```
foma[0]: source g2p.foma
Opening file 'g2p.foma'.
...
192.0 kB. 260 states, 12165 arcs, Cyclic.
foma[1]: down
apply down> alabama
a0la0ba1ma0
```

So the result is as expected, stress on second vowel from the end of word.
Prepared test.ssml with following 3 lines:

```
<speak xml:lang="pl">
<s>alabama</s>
</speak>
```

Then I called:

```
RHVoice-transcribe-sentences -s test.ssml test.txt
```

and got:

```
pau a1 l a1 b a1 m a1 pau 
```

So, unexpected stress on each vowel.

I checked out in the repo to the state, where Polish was defined in code, and everything worked fine there:

```
pau a l a b a1 m a pau 
```

Since converting Polish to data_only, only g2p was modified / improved, not other foma files.

The stress adding in foma is as follows:

```
define procStress1 [a:a1|
on:on1|
e:e1|
en:en1|
i:i1|
o:o1|
u:u1|
ih:ih1
] @-> ||
_ \VowelPhonem+ VowelPhonem+ .#.,
_ \VowelPhonem+ VowelPhonem+ \VowelPhonem+ .#.,
.#. \VowelPhonem+ _ \VowelPhonem+ .#.;

define procStress2 [a:a0|
on:on0|
e:e0|
en:en0|
i:i0|
o:o0|
u:u0|
ih:ih0
] @-> || _ ;


regex Stage1 .o.
processMultigraphs1 .o.
processMultigraphs2 .o.
processMultigraphs3 .o.
Stage2 .o.
procStress1 .o. procStress2;
```

**Reply by Olga-Yakovleva** [reply]
Just an idea without looking at any code: did you correct your syl.foma to include vowels with stress?

**Comment by zstanecic** [comment]
Hi Olga,

Thanks! We don’t have this declared in the syllabification code.

**Comment by grzezlo** [comment]
As I supposed before: mysterious magic, that's all.
I would never thought, that syl.foma has something to do about changing multiple "a0" as outputted by g2p, to "a1" printed out by transcribe-sentences.
But in fact, it had, and updating syl solved the problem.
Thanks.
It restored intonation for exclamation and question sentences in Polish language.
I think this issue can be closed.
Thanks again for your help, @louderpages and @Olga-Yakovleva.

**Comment by zstanecic** [comment]
Yep, it can be closed.

I am now retraining prototypes.
